spec/integration/parser/function_parameters_spec.rb

```diff
@@ -0,0 +1,94 @@
+require 'spec_helper'
+require 'puppet_spec/language'
+
+module PuppetSpec
+module FunctionRunner
+  extend RSpec::Matchers::DSL
+
+  def self.collect_log(decl, call)
+    Puppet[:code] = decl
+    compiler = Puppet::Parser::Compiler.new(Puppet::Node.new('specification'))
+    evaluator = Puppet::Pops::Parser::EvaluatingParser.new()
+    logs = []
+    Puppet::Util::Log.with_destination(Puppet::Test::LogCollector.new(logs)) do
+      compiler.compile
+      evaluator.evaluate_string(compiler.topscope, call)
+    end
+    logs = logs.select { |log| log.level == :notice }.map { |log| log.message }
+    logs
+  end
+
+  def run_expectations(decl, expectations)
+    expectations.each do |call, result|
+      if result.is_a?(Regexp)
+        it "'#{decl}' called with '#{call}' should fail with error #{result}" do
+          expect {FunctionRunner.collect_log(decl, call)}.to raise_error(Puppet::Error, result)
+        end
+      else
+        it "'#{decl}' called with '#{call}' should produce #{result}" do
+          expect(FunctionRunner.collect_log(decl, call)).to include(*result)
+        end
+      end
+    end
+  end
+end
+
+describe "Function parameter default expressions" do
+  extend FunctionRunner
+
+  context "to the left of a parameter can use that parameter's value" do
+    run_expectations('function example($a = 10, $b = $a) { notice("\$a == ${a}") notice("\$b == ${b}") }',
+      { 'example()' => ['$a == 10', '$b == 10'],
+        'example(2)' => ['$a == 2', '$b == 2'],
+        'example(2,5)' => ['$a == 2', '$b == 5'] })
+
+    run_expectations('function example($a, $b = $a) { notice("\$a == ${a}") notice("\$b == ${b}") }',
+      { 'example(10)' => ['$a == 10', '$b == 10'],
+        'example(10,20)' => ['$a == 10', '$b == 20'] })
+  end
+
+  context 'that references a variable to the right' do
+    run_expectations('function example($a = 10, $b = $c, $c = 20) { notice("\$a == ${a}") notice("\$b == ${b}") notice("\$c == ${c}") }',
```

Comment:
need to break the long lines, use a heredoc to make it easier to read the puppet source ?